### PR TITLE
[CAPPL-222] feat(workflows): adds a secrets syncer for workflow registry

### PR DESCRIPTION
**Pull request by MStreet3**
[CAPPL-222](https://smartcontract-it.atlassian.net/browse/CAPPL-222)

See the design doc [here](https://smartcontract-it.atlassian.net/wiki/spaces/CAP/pages/1093435476/Workflow+Registry+Syncer+Design).

This PR:
* implements a secrets sync in the workflow registry syncer
* removes `indexed` modifier from the WorkflowRegistry's `WorkflowRegistryRequestedForceUpdateSecretsV1` event
* adds a `workflow_secrets` DB table and `ORM` interface for interacting with it
* functional test against a simulated EVM backend in EVM relays

### Requires

#14990 

### Supports

[CAPPL-222]: https://smartcontract-it.atlassian.net/browse/CAPPL-222?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: contracts/scripts/native_solc_compile_all_workflow [hunk 1]
```diff
@@ -11,12 +11,12 @@ OPTIMIZE_RUNS=1000000
...
-python3 -m pip install --require-hashes -r "$SCRIPTPATH"/requirements.txt
+# pip3 install --upgrade pip
+# python3 -m pip install --require-hashes -r "$SCRIPTPATH"/requirements.txt
 
-if ! solc-select versions | grep -q "$SOLC_VERSION"; then
-  solc-select install $SOLC_VERSION
-fi
+# if ! solc-select versions | grep -q "$SOLC_VERSION"; then
+  # solc-select install $SOLC_VERSION
+# fi
```
Comment (cedric-cordenier) [inline contracts/scripts/native_solc_compile_all_workflow]: @MStreet3 I'm guessing this is unintentional :) ?

File: core/services/workflows/secrets/worker.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+func (w *worker) Close() error {
+	return w.StopOnce("secrets_worker", func() error {
+		close(w.stopCh)
+		w.wg.Wait()
+		return nil
+	})
+}
+
+func (w *worker) getTicker(ctx context.Context) <-chan struct{} {
```
Comment (cedric-cordenier) [inline core/services/workflows/secrets/worker.go]: @MStreet3 Why is this needed as opposed to something like:

```
timer := s.newTimer()
defer timer.Close()
```

in the main loop?

File: core/services/workflows/syncer/workflow_registry.go [hunk 3]
```diff
@@ -2,39 +2,182 @@ package syncer
...
+
+func (w *workflowRegistry) syncLoop() {
+	ctx, cancel := w.stopCh.NewCtx()
+	defer cancel()
+
+	ticker := w.getTicker(ctx)
+
+	// Sync for a first time outside the loop; this means we'll check the cached state
+	// or start a remote sync immediately once spinning up syncLoop, as by default a ticker will
+	// fire for the first time at T+N, where N is the interval.
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/workflow_registry.go]: Note: you could use chainlink-common/pkg/timeutil/ticker here with nextDur returning 0 the first time it's called

File: core/services/workflows/syncer/workflow_registry.go [hunk 4]
```diff
@@ -2,39 +2,182 @@ package syncer
...
+			err := w.Sync(ctx, false)
+			if err != nil {
+				w.lggr.Errorw("failed to sync with remote registry", "error", err)
+				w.sinkError(err)
+			}
+		}
+	}
+}
+
+func (w *workflowRegistry) updateStateLoop() {
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/workflow_registry.go]: I'd recommend using a stateless syncer initially to reduce complexity. Stateless things have much less change of going wrong.

File: core/services/workflows/syncer/workflow_registry.go [hunk 5]
```diff
@@ -2,39 +2,182 @@ package syncer
...
+				w.lggr.Errorw("failed to save state to local registry", "error", err)
+				w.sinkError(err)
+			}
+		}
+	}
+}
+
+// sinkError is a non-blocking attempt to sink errors to the registry syncer's error channel.
+// If no one reads, the errors are dropped.
+func (w *workflowRegistry) sinkError(err error) {
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/workflow_registry.go]: It looks like this is primarily for testing: if so can we add that to the comments?

File: core/store/migrate/migrations/0259_add_workflow_artifacts.sql [hunk 6]
```diff
@@ -0,0 +1,43 @@
+-- +goose Up
+-- +goose StatementBegin
+-- Create the workflow_artifacts table
+CREATE TABLE workflow_artifacts (
```
Comment (cedric-cordenier) [inline core/store/migrate/migrations/0259_add_workflow_artifacts.sql]: Might as well call this workflow_secrets since it'll only be used to store secrets

File: core/store/migrate/migrations/0259_add_workflow_artifacts.sql [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ADD COLUMN binary_url TEXT,
+ADD COLUMN artifact_id INT UNIQUE REFERENCES workflow_artifacts(id) ON DELETE CASCADE;
+
+-- Alter the config column type
+ALTER TABLE workflow_specs
+ALTER COLUMN config TYPE TEXT;
+
+-- Set the default value for the config column
+ALTER TABLE workflow_specs
+ALTER COLUMN config SET DEFAULT '';
```
Comment (cedric-cordenier) [inline core/store/migrate/migrations/0259_add_workflow_artifacts.sql]: My PSQL knowledge is a bit rusty, but adding a default (as opposed to leaving it nullable) will require a lock on the entire table + the modification of each row. Depending on how large the table is that could be more or less dangerous.

In this case I think it's fine since we're only running this on staging and the tables are small.

File: core/store/migrate/migrations/0259_add_workflow_artifacts.sql [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    secrets_url_hash TEXT UNIQUE,
+    contents BYTEA
+);
+
+-- Create an index on the secrets_url_hash column
+CREATE INDEX idx_secrets_url_hash ON workflow_artifacts(secrets_url_hash);
+
+-- Alter the workflow_specs table
+ALTER TABLE workflow_specs
+ADD COLUMN binary_url TEXT,
```
Comment (cedric-cordenier) [inline core/store/migrate/migrations/0259_add_workflow_artifacts.sql]: Should we also add a config_url for completeness?

File: core/services/workflows/secrets/worker.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+	if w.timer == nil {
+		w.timer = signalers.MakeTicker(ctx.Done(), defaultTickInterval)
+	}
+	return w.timer
+}
+
+// syncForceUpdateSecretsEvents synchronizes the force update secrets events from the contract
+// to the local database.
+func (w *worker) syncForceUpdateSecretsEvents(
```
Comment (cedric-cordenier) [inline core/services/workflows/secrets/worker.go]: @MStreet3 YAGNI the concurrency here: it's abstracting things away thus making things a bit harder to understand and I'm not sure what we're gaining here.

This operation should correspond to:
- updating the relevant secrets entry in the DB
- updating the local in-mem cache for the secrets

I'd also recommend combining this package with the syncer package for more cohesion. There's no need for a separate worker here: just parse the event and call the handler, potentially mediating by a worker pool if you need that parallelism that gives you.

File: core/services/workflows/secrets/handler.go [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+
+					var dm map[string]any
+					err = vm.UnwrapTo(&dm)
+					if err != nil {
+						sendErr(err)
+						continue
+					}
+
+					var event T
+					err = mapstructure.Decode(dm, &event)
```
Comment (cedric-cordenier) [inline core/services/workflows/secrets/handler.go]: @MStreet3 Hmmmm... As I'm reading this: you're dumping something to a values.Map (WrapMap), then unwrapping it to a map[string]any, then unwrapping it to a struct. Can we somehow just dump it straight to the struct?

File: core/services/workflows/secrets/handler.go [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+func newForceUpdateSecretsHandler(orm ORM, fetcher FetcherFunc) Handler {
+	return &handler{orm: orm, fetcher: fetcher}
+}
+
+func (h *handler) ForceUpdateSecrets(
+	ctx context.Context,
+	event Event,
+) error {
+	// Fetch the secrets url from ORM
+	url, err := h.orm.GetSecretsURL(ctx, event.GetSecretsURL())
```
Comment (cedric-cordenier) [inline core/services/workflows/secrets/handler.go]: Is this correct ? You're fetching the secrets URL from the event, and then are getting it from the DB?

**Comment by cedric-cordenier** [comment]
@MStreet3 Great first stab! Thanks for turning this around quickly.

In terms of high-level feedback, I'd recommend you simplify the implementation: there's a lot here that is increasing implementation cost and abstraction overhead without a corresponding benefit. You shouldn't need much more than:
- a polling loop -> which returns the events we care about
- a worker loop -> which receives the events and processes them

This should all be located in a single package; we can say at this stage that they won't need to be reused. (In contrast to the registrysyncer + launcher construct; this is separate because we have two different launchers today -- something which we won't have for the workflow syncer.)

If you want to do the work faster you can have a worker pool of some kind, but note you'll need to be careful that we preserve the order of the events when processing (you could send all messages for a workflow to the same worker as a way of solving this). I'm not sure we need that at this stage though.

File: contracts/scripts/native_solc_compile_all_workflow (hunk 1 above)
Comment (MStreet3) [reply contracts/scripts/native_solc_compile_all_workflow]: Yea was a hack to fix a compilation issue

File: core/services/workflows/secrets/worker.go (hunk 2 above)
Comment (MStreet3) [reply core/services/workflows/secrets/worker.go]: This is for letting tests override using a ticker for a channel instead. Removes timing from unit tests.

Comment (cedric-cordenier) [reply core/services/workflows/secrets/worker.go]: Yep, understood -- I was commenting on why we need to lazy load (as opposed to dependency inject the dependency) and use a signaler

File: core/services/workflows/secrets/handler.go (hunk 11 above)
Comment (MStreet3) [reply core/services/workflows/secrets/handler.go]: The event only has the hash, so get the plain url from the table, then fetch the contents, then write

File: core/services/workflows/secrets/handler.go (hunk 10 above)
Comment (MStreet3) [reply core/services/workflows/secrets/handler.go]: Yea this is mostly copied from log event trigger, will take a look again

File: core/services/workflows/secrets/worker.go (hunk 9 above)
Comment (MStreet3) [reply core/services/workflows/secrets/worker.go]: Hm, there's no extra concurrency here. This method just starts the secrets event query worker and passes the events to the event handler.

File: core/services/workflows/syncer/workflow_registry.go (hunk 4 above)
Comment (MStreet3) [reply core/services/workflows/syncer/workflow_registry.go]: Removed. Was scratch code

File: core/services/workflows/syncer/handler.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
+package syncer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+type Handler interface {
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/handler.go]: Nit/can follow up: does this need to be public? Same with the interface method?

File: core/services/workflows/syncer/handler.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+	gateway FetcherFunc,
+) *eventHandler {
+	return &eventHandler{
+		handlers: map[WorkflowRegistryEventType]Handler{
+			ForceUpdateSecretsEvent: newForceUpdateSecretsHandler(lggr, orm, gateway),
+		},
+	}
+}
+
+func (h *eventHandler) Handle(ctx context.Context, event WorkflowRegistryEvent) error {
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/handler.go:32]: Same with this

File: core/services/workflows/syncer/handler.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+	}
+}
+
+// Handle processes the ForceUpdateSecretsEvent by fetching the secrets from the URL for a given event
+// and updating the local state.
+func (h *forceUpdateSecretsHandler) Handle(
+	ctx context.Context,
+	event WorkflowRegistryEvent,
+) error {
+	if event.EventType != ForceUpdateSecretsEvent {
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/handler.go]: This should be moved higher up: I'd expect we'd basically do this same logic just above this, in something like, so this would effectively be duplicated

```
switch event.EventType {
case ForceUpdateSecretsEvent:
    secretsHandler.Handle(ctx, event)
...
```

File: core/services/workflows/syncer/handler.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	// Fetch the contents of the secrets file from the url via the fetcher
+	secrets, err := h.fetcher(ctx, url)
+	if err != nil {
+		return err
+	}
+
+	// Update the secrets in the ORM
+	if _, err := h.orm.Update(ctx, url, string(secrets)); err != nil {
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/handler.go]: You're going back and forth here a bit: you have bytes, are converting it to a string, then convert it to bytes, and base64 encode it.

Why not just store it as a bytea, or pass in bytes and base64 encode it before dumping it in the db?

File: core/services/workflows/syncer/orm.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+	"github.com/smartcontractkit/chainlink-common/pkg/sqlutil"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+type ORM interface {
+	// GetSecretsURLByID returns the secrets URL for the given ID.
+	GetSecretsURLByID(ctx context.Context, id int64) (string, error)
+
+	// GetContents returns the contents of the secret at the given plain URL.
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/orm.go:18]: Nit: secret -> secrets (it has more than one secret inside it)

File: core/services/workflows/syncer/orm.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+	return &orm{
+		ds:   ds,
+		lggr: lggr,
+	}
+}
+
+func (orm *orm) GetSecretsURLByID(ctx context.Context, id int64) (string, error) {
+	var secretsURL string
+	if err := orm.ds.GetContext(ctx, &secretsURL,
+		`SELECT secrets_url FROM workflow_secrets WHERE workflow_secrets.id = $1`,
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/orm.go:52]: Do we need to base64 encode the secrets_url? Why not store it in plaintext in the DB?

File: core/services/workflows/syncer/orm.go [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+	).Scan(&id)
+
+	if err != nil {
+		return 0, err
+	}
+
+	return id, nil
+}
+
+func (orm *orm) SecretsFor(ctx context.Context, workflowOwner, workflowName string) (map[string]string, error) {
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/orm.go]: Are you planning on filling this method in as part of this PR?

Depending on what this does we may want to move it out of the ORM, since we need to decrypt the secrets, perform validation on it etc.

File: core/services/workflows/syncer/workflow_registry.go [hunk 19]
```diff
@@ -2,39 +2,366 @@ package syncer
...
+
+var (
+	// ForceUpdateSecretsEvent is emitted when a request to force update a workflows secrets is made
+	ForceUpdateSecretsEvent WorkflowRegistryEventType = "WorkflowForceUpdateSecretsRequestedV1"
+)
+
+// WorkflowRegistryEvent is an event emitted by the WorkflowRegistry.  Each event is typed
+// so that the consumer can determine how to handle the event.
+type WorkflowRegistryEvent struct {
+	logeventcap.Output
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/workflow_registry.go]: You shouldn't need to depend on the logevent capability; so this feels odd. What made you add this?

File: core/services/workflows/syncer/workflow_registry.go [hunk 20]
```diff
@@ -2,39 +2,366 @@ package syncer
...
+		opt(wr)
+	}
+	return wr
+}
+
+// Start starts the workflowRegistry.  It starts two goroutines, one for querying the contract
+// and one for handling the events.
+func (w *workflowRegistry) Start(_ context.Context) error {
+	return w.StartOnce(w.Name(), func() error {
+		ctx, _ := w.stopCh.NewCtx()
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/workflow_registry.go]: You can't ignore the cancel here; that's a goroutine that will leak.

You can avoid this by creating separate contexts inside each goroutine, eg.

```
ctx, cancel := w.stopCh.NewCtx()
defer cancel()
defer w.wg.Done()
w.queryLoop(ctx, w.cfg)
```

File: core/services/workflows/syncer/workflow_registry.go [hunk 21]
```diff
@@ -2,39 +2,366 @@ package syncer
...
+		select {
+		case <-ctx.Done():
+			return
+		case resp, open := <-w.eventsCh:
+			if !open {
+				return
+			}
+
+			if resp.Err != nil || resp.Event == nil {
+				continue
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/workflow_registry.go:256]: You could log the err here; atm we would have no way of knowing that an error occurred

File: core/services/workflows/syncer/workflow_registry.go (hunk 19 above)
Comment (cedric-cordenier) [reply core/services/workflows/syncer/workflow_registry.go]: Ah you're just using the struct: can we copy it instead? There really shouldn't be any dependencies between these two components; if this happens again we should consider asking the chain reader team if we can upstream it somehow.

File: core/store/migrate/migrations/0259_add_workflow_secrets.sql [hunk 22]
```diff
@@ -0,0 +1,40 @@
+-- +goose Up
+-- +goose StatementBegin
+-- Create the workflow_artifacts table
+CREATE TABLE workflow_secrets (
+    id SERIAL PRIMARY KEY,
+    secrets_url TEXT UNIQUE, -- base64 encoded URL
```
Comment (cedric-cordenier) [inline core/store/migrate/migrations/0259_add_workflow_secrets.sql]: See my comment about whether this can just be stored in plaintext

As an aside, one thing that jump out here is that different URLs can point to the same resource, so we may need to normalize the URLs at some point; I think it's fine to leave for now though.

File: core/services/workflows/syncer/handler.go (hunk 12 above)
Comment (MStreet3) [reply core/services/workflows/syncer/handler.go]: Yea unnecessarily public.

File: core/services/workflows/syncer/workflow_registry.go [hunk 23]
```diff
@@ -2,39 +2,557 @@ package syncer
...
+// WorkflowRegistryEventResponse is a response to either parsing a queried event or handling the event.
+type WorkflowRegistryEventResponse struct {
+	Err   error
+	Event *WorkflowRegistryEvent
+}
+
+// ContractReader is the subset of methods needed to query the events of a contract.
+type ContractReader interface {
+	Bind(context.Context, []types.BoundContract) error
+	QueryKey(
```
Comment (ilija42) [inline core/services/workflows/syncer/workflow_registry.go:71]: Why is there a need for this? If you were calling the go eth client, would you make a wrapper each time for the functionality that you are using? I just want to discourage this so that we don't make interface changes and versioning harder for no reason

File: core/services/workflows/syncer/heap.go [hunk 24]
```diff
@@ -0,0 +1,63 @@
+package syncer
+
+import "container/heap"
+
+type Heap interface {
```
Comment (cedric-cordenier) [inline core/services/workflows/syncer/heap.go:5]: Future style note: it's best to abstract as late as possible and to avoid altogether if possible; in this case you have exactly one implementation of the heap (blockHeighHeap), so you could have avoided declaring these at all (I think anyway -- I'm not sure I follow why publicHeap is necessary).

Comment (cedric-cordenier) [reply core/services/workflows/syncer/heap.go:5]: I wouldn't expect the public thing to come into it at all on reflection: this is an internal implementation detail of the syncer, so shouldn't be exposed publicly. I may well be missing something though.

**Review by cedric-cordenier — DISMISSED** [review]
